File: src/solution_java/medium/_53MaximumSubarray.java
package solution_java.medium;

public class _53MaximumSubarray {
    public class Solution {
        public int maxSubArray(int[] nums) {
            int n = nums.length;
            if (n == 1) {
                return nums[0];
            }
            int result = nums[0];
            int[] dp = new int[n];
            dp[0] = result;
            for (int i = 1; i < n; i++) {
                dp[i] = Math.max(dp[i - 1], 0) + nums[i];
                result = Math.max(result, dp[i]);
            }
            return result;
        }
    }

    public class BaseSolution {

        public int maxSubArray(int[] nums) {
            int pre = 0;
            int res = nums[0];
            for (int num : nums) {
                pre = Math.max(pre + num, num);
                res = Math.max(res, pre);
            }
            return res;
        }
    }
}

File: src/easy/_53MaximumSubarray.kt
package easy

class _53MaximumSubarray {
    class Solution {
        fun maxSubArray(nums: IntArray): Int {
            val dp = IntArray(nums.size)
            var max = Integer.MIN_VALUE
            for (index in nums.indices) {
                if (index == 0 || dp[index - 1] <= 0) {
                    dp[index] = nums[index]
                } else {
                    dp[index] = dp[index - 1] + nums[index]
                }
                max = Math.max(max, dp[index])
            }
            return max
        }
    }

    class BestSolution {
        fun maxSubArray(nums: IntArray): Int {
            val n = nums.size
            val dp = IntArray(n)
            dp[0] = nums[0]
            var result = dp[0]
            for (i in 1 until n) {
                dp[i] = Math.max(dp[i - 1], 0) + nums[i]
                result = Math.max(result, dp[i])
            }
            return result
        }
    }
}